.github/workflows/test.yml

```diff
@@ -6,10 +6,36 @@ on:
     types: [opened, synchronize]
     branches-ignore: [staging, production]
 
+env:
+  # Number of parallel jobs for jest tests
+  CHUNKS: 3
 jobs:
+  config:
+    runs-on: ubuntu-latest
+    name: Define matrix parameters
+    outputs:
+      MATRIX: ${{ steps.set-matrix.outputs.MATRIX }}
+    steps:
+      - name: Set Matrix
+        id: set-matrix
+        uses: actions/github-script@v6
+        with:
+          # Generate matrix array i.e. [0, 1, 2, ...., CHUNKS] for test job
+          script: |
+            core.setOutput('MATRIX', Array.from({ length: process.env.CHUNKS + 1 }, (v, i) => i));
```

Comment:
I see that we are using this run to define a number of chunks which seems like completely unnecessary work in GitHub runner. 

Is there a way to get the number of matrices from the configuration in jest or somewhere in test files without this step?

Reply:
We are doing this to reduce time in CI so it make sense to add here in respective action. Also, there is no such configuration in jest to run test in chunks and define some parameter.

Reply:
> We are doing this to reduce time in CI so it make sense to add here in respective action

I agree but the problem with this is that it will be fragile. As soon as someone modifies a test file and a time-consuming test case to it, it will start consuming more time. 


Now, let's say that the user wants to optimize the execution time of tests. what are his options? 
1. Modify the test files order so that time-consuming tests are grouped in separate chunck. 
2. Break tests in subsets. 
4. Modify the action file to increase the chunks. 

I don't think either of these is very ideal. It is overhead. 

Instead, if there is a way to configure some kind of groups and these run parallel in its matrix then the user will just change the configuration and it will all be handled automatically. 

One way of achieving this could be the use of groups. https://www.npmjs.com/package/jest-runner-groups. 

Assign each test to a group. While running the tests via CI get the number of groups and create a matrix of length equal to # groups. 

What do you thing/

Reply:
But first, let's get rid of the `Set Matrix` step and use a static matrix. I don't see any advantage of this step. 

The way user is setting `chunk: n` env, they can update the matrix : [0, 2, 3, ..n].

Reply:
What are your thoughts?

Reply:
> But first, let's get rid of the `Set Matrix` step and use a static matrix. I don't see any advantage of this step.
> 
> The way user is setting `chunk: n` env, they can update the matrix : [0, 2, 3, ..n].

@parasharrajat  Only advantage to use this steps and common `env.CHUNKS` is that we don't have magic numbers all over and f someone forgets to update those numbers in commands and just updates the matrix, it will break/fail.

Reply:
> > We are doing this to reduce time in CI so it make sense to add here in respective action
> 
> I agree but the problem with this is that it will be fragile. As soon as someone modifies a test file and a time-consuming test case to it, it will start consuming more time.
> 
> Now, let's say that the user wants to optimize the execution time of tests. what are his options?
> 
> 1. Modify the test files order so that time-consuming tests are grouped in separate chunk.
> 2. Break tests in subsets.
> 3. Modify the action file to increase the chunks.
> 
> I don't think either of these is very ideal. It is overhead.
> 
> Instead, if there is a way to configure some kind of groups and these run parallel in its matrix then the user will just change the configuration and it will all be handled automatically.
> 
> One way of achieving this could be the use of groups. https://www.npmjs.com/package/jest-runner-groups.
> 
> Assign each test to a group. While running the tests via CI get the number of groups and create a matrix of length equal to # groups.
> 
> What do you thing?

We can do this but it has more maintainance cost as developer need to put new test files in specific group or new groups. They have to first check which group is taking most time and then adjust it by changing group name.

Also, using this package, one can assign more than one group (intentionally or unintentionally) to single test file, causing same file being run/executes by multiple runners.

But if you strongly feel it's better than what we have, I can do that.

Reply:
Ok. Thanks. I think we can drop this for now.

Reply:
@parasharrajat  If everything is good, can you kindly add your review checklist so all steps go green.